04_Scheduling/02_Scheduler.md

```diff
@@ -0,0 +1,149 @@
+        break;
+    case TIMER_INTERRUPT:
+        // eventually doing some other stuff 
+        schedule(); // <-- here we call the scheduler
+        break;
+    case ANOTHER_INTERRUPT:
+        // ...
+        break;
+}
+```
+
+Is that all? More or less, we may eventually want to have it called in other cases too (i.e. maybe while serving a syscall, or create a custom interrupt to handle specific cases), but the logic is always the same, we call it in all the parts where we decide is time to have a task switch. 
+
+### Part 2 Checking if the task should be preempted
+
+Even if we are not really doing a check in our simple algorithm, because we will have it switched at every `schedule()` call, is worth spending few words about this step. 
+
+Depending on the algorithm selected (or created) we can decide to let tasks execute for more time than one *quantum* (this terms indicates the amount of time between two ticks of the timer), and in this case we will need to implement a mechanism to decidere whether the task should be interrupted or not. 
+
+When the function is called we need to check if it has finished it's allocated time. Who decides it? How long it is? How we calculate it? Well the answer is that this is a design choice, we can schedule a task at every single timer interrupt, or give it a certain number of *ticks*, that number can be fixed (decided at compile time, or by a configuration parameter of the kernel), or variable (for example if we are having tasks with different priorities, maybe we want to give more time to higher priority tasks). But in any case the minimum amount of time for a task in execution is for at least 1 *tick*. 
+
+As mentioned above we will try to keep things simple and will change task every time the scheduler is called. So we will skip this check
+
+### Next stuff to be completed...
+
+Now if the `current_thread` har reached it's allocated time, it's time to pick the next one. But before doing that we need to save make sure that next time `current_thread` will be picked up, it will resume from the exact point it is being interrupted. This is achieved by saving the current execution context. And what is it? Well we already encountered that, in the [interrupt handling](../InterruptHandling.md) chapter, it is the status of the cpu in that exact istant, all the registers value (the instruction pointer, the stack values, the general purpose registers etc). And this gives us a big hint on how we are going to switch between threads: we will avail of our interrupt handler (*Authors note: of course this is not the only way to switch between task, but this is in our opinion one of the easiest to implement).*)
+
+Let's recall quickly what our interrupt handling routine does: 
+
+* It first save the context pushing all registers on the stack
+* Then it calls the interrupt handling function, that serves the interrupt
+* It restore the context from the stack
+* Return (if possible) the control to the kernel
```

Comment:
I see why you've put the 'if possible', but I think it'd be more accurate to rewrord this. Since we're returning to whatever the iret frame indicates (could be user code, could be user code, could even be another interrupt handler if we support nested interrupts). Perhaps word it like:

Returns control to the previously interrupted code.